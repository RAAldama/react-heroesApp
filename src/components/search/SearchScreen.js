import React, { useMemo } from "react";
import queryString from "query-string";
import { useLocation } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import HeroCard from "../heroes/HeroCard";
import { getHeroByName } from "../../selectors/getHeroByName";

const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({ hero: q });
  const { hero } = formValues;

  const heroesFiltered = useMemo(() => getHeroByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    history.push(`?q=${hero}`);
  };

  return (
    <div>
      <h1>Search!</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Busqueda</h4>
          <hr />

          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder0="Search a hero!"
              className="form-control"
              name="hero"
              value={hero}
              onChange={handleInputChange}
            />

            <div className="d-grid">
              <button
                type="submit"
                className="btn btn-outline-info mt-1 btn-block"
              >
                Agregar
              </button>
            </div>
          </form>
        </div>

        <div className="col-7">
          <h4>Heroes encontrados</h4>
          <hr />

          {q === "" && (
            <div className="alert alert-info">Â¡Busca tu heroe favorito!</div>
          )}

        {(q !== "" && heroesFiltered.length === 0) && (
            <div className="alert alert-danger">No existe un heroe en la base con el nombre {q}</div>
          )}

          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
