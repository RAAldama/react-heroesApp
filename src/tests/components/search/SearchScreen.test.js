import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router-dom"
import SearchScreen from "../../../components/search/SearchScreen"

describe('tests on <SearchScreen />', () => {
    
    test('should render correctly with default values', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Â¡Busca tu heroe favorito!');
    })

    test('should show superman and the input with the query', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=superman']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe('superman');
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should show a message of error if there is no hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=xd']}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-danger').exists() ).toBe(true);
        expect( wrapper.find('.alert-danger').text().trim() ).toBe('No existe un heroe en la base con el nombre xd');
    })
    
    test('should call the push of history', () => {
        const history = {
            push: jest.fn(),
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=xd']}>
                <Route path="/search" component={() => <SearchScreen history={history}/>} />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'hero',
                value: 'superman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(history.push).toHaveBeenCalledWith('?q=superman');
    })
    
})
