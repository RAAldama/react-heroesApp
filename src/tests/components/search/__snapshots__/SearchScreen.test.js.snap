// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <SearchScreen /> should render correctly with default values 1`] = `
<div>
  <h1>
    Search!
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Busqueda
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="hero"
          onChange={[Function]}
          placeholder0="Search a hero!"
          type="text"
          value=""
        />
        <div
          className="d-grid"
        >
          <button
            className="btn btn-outline-info mt-1 btn-block"
            type="submit"
          >
            Agregar
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Heroes encontrados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        ¡Busca tu heroe favorito!
      </div>
    </div>
  </div>
</div>
`;

exports[`tests on <SearchScreen /> should show superman and the input with the query 1`] = `
<div>
  <h1>
    Search!
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Busqueda
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="hero"
          onChange={[Function]}
          placeholder0="Search a hero!"
          type="text"
          value="superman"
        />
        <div
          className="d-grid"
        >
          <button
            className="btn btn-outline-info mt-1 btn-block"
            type="submit"
          >
            Agregar
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Heroes encontrados
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <img
            alt="Superman"
            className="card-img-top"
            src="./assets/dc-superman.jpg"
          />
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Superman
            </h5>
            <p
              className="card-text"
            >
              Kal-El
            </p>
            <a
              href="/hero/dc-superman"
              onClick={[Function]}
            >
              Más información del héroe...
            </a>
          </div>
          <div
            className="card-footer"
          >
            <small
              className="text-muted"
            >
              Action Comics #1
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
